### PR TITLE
[history] update to 4.5.0

### PR DESCRIPTION
**Pull request by rokoroku**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Run `tsc` without errors.
- [ ] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/mjackson/history
- [x] Increase the version number in the header if appropriate.

Current definition for history is too outdated. (There are lots of breaking changes! )
Update it to follow v4.5.0.

### CODE REVIEW

**Comment by Gaelan** [comment]
@andy-ms Not the author, but builds are now passing. Mergable now?

**Comment by rokoroku** [comment]
Good, restored legacy files to the ``v2`` folder.

**Comment by andy-ms** [comment]
So `history-2.0.d.ts` is now `history/v2/index.d.ts`? Can you remove `history-2.0.d.ts`?

**Comment by rokoroku** [comment]
@andy-ms Not really. Currently ``history-2.0.d.ts`` is still required for backward-compatibility issues. It's not same with ``history/v2/index.d.ts``.

Should I merge the backward-compat declarations into ``index.d.ts``?

**Comment by rokoroku** [comment]
See the last **travis-ci** error logs below line 1501: 
https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/187267426#L1501

And see ``react-router`` declaration files:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/react-router/index.d.ts#L6

``history`` v4 no more provide ``Query`` interfaces in any place, but other dependencies (using ``history`` below v3) requires it. 
Until these coupling is removed, backward compatibility is required I think.

**Comment by andy-ms** [comment]
The solution would be either to upgrade those, or to give them a [package.json](https://github.com/DefinitelyTyped/DefinitelyTyped#i-notice-some-packages-having-a-packagejson-here) depending on `"@types/history": "^2.0.42"` until they can be upgraded.

**Comment by rokoroku** [comment]
@andy-ms  I Tried adding  **package.json** with ``@types/history": "^2.0.42"`` to the ``react-router`` package, but because of ``baseUrl: "../"`` from **tsconfig.json**, ``import from 'history'`` still resolved from ``../history`` which is rewritten for v4 .

So I removed baseUrl property and add ``"@types/react": "^0.14.55"`` as a new dependency (because it requires it). Then, ``react-router`` passed the test. However, other packages using ``react-router`` or ``history`` (directly from ``baseUrl``) still make same error.

The related packages are:
```
Error in react-breadcrumbs
Error in react-i18next
Error in react-router-bootstrap
Error in react-router-redux
Error in redux-bootstrap
Error in redux-router
```

So, in this approach, I need to do same works to all the related packages ... am I right?

**Comment by andy-ms** [comment]
Try doing this in those packages:
```json
"typeRoots": [
    "node_modules",
    "../"
],
```

**Comment by rokoroku** [comment]
Well, I asked the problem with ``"baseUrl": "../"``. Without removing it, resolving non-relative module (``import from 'history'``) always resolved from ``../history`` directory, not from ``node_modules/@types/history``.

So now I changed all those ``history``-dependent module's ``package.json`` and ``tsconfig.json`` without **baseUrl** settings, omitting backward compat declarations from ``history``.

Is this all right?

File: package.json [hunk 1]
```diff
@@ -23,5 +23,10 @@
   },
   "devDependencies": {
     "types-publisher": "Microsoft/types-publisher#production"
+  },
```
Comment (andy-ms) [inline package.json]: This is the root `package.json`, it should not be changed.

**Comment by andy-ms** [comment]
Sorry for the delay. Removing `baseUrl` should be fine. But if you are depending on the *latest* version of a package, don't add it to `package.json`, because then they would be tested against the one on NPM and not the one available locally. (Meaning, if we're about to publish something which breaks tests, we won't get test failures until we've already published it.) Use a [path mapping](http://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping) instead.

**Comment by rokoroku** [comment]
Since this PR is so messed, I just made a new PR. see #13744

**Pull request closed without merging**